### PR TITLE
Exec API (Experimental)

### PR DESCRIPTION
**Pull request by Conner1115**
Add the experimental `exec` API.

## Usage

```ts
import { exec } from '@replit/extensions';

...

exec(args)
```

## Call Signature
```ts
exec(args: ExecArgs) => Promise<() => void>;
```

## New Types
```ts
/**
 * The promise result returned by the `exec()` function
 */
export interface ExecResult {
  exitCode: number | null;
  error: string | null;
}

/**
 * Custom environment variables for an execution channel
 */
export type Env = {
  [key: string]: string;
};

/**
 * Options for the `exec()` function.
 * `splitStderr` - Whether to seperate stderr from stdout
 * `env` - Optional. Whether to add custom environment variables to this execution channel
 * `args` - The arguments to execute.  Can be a string or an array of strings.
 * `onOutput` - Callback for when the output of the command is streamed.
 * `onStdErr` - Callback for when the stderr of the command is streamed.
 * `onError` - Callback for when the command errors.
 * `onEnd` - Callback for when the command ends.
 */
export interface ExecArgs {
  splitStderr?: boolean;
  env?: Env;
  args: Array<string> | string;
  onOutput?: (output: string) => void;
  onStdErr?: (stderr: string) => void;
  onError?: (error: Error) => void;
  onEnd?: (
    res: ExecResult & {
      output: string;
    }
  ) => void;
}

/**
 * Options for the replit-web exec function
 */
export interface ExecOptions {
  splitStderr?: boolean;
  args: Array<string> | string;
  env?: Env;
  onOutput?: (output: string) => void;
  onStdErr?: (stderr: string) => void;
  onError?: (error: Error) => void;
}
```

### CODE REVIEW

File: modules/extensions/src/api/exec.ts [hunk 1]
```diff
@@ -0,0 +1,59 @@
...
+}
+
+export default async function exec({
+  splitStderr,
+  args,
+  env,
+  onOutput = () => {},
+  onStdErr = () => {},
+  onError = () => {},
+  onEnd = () => {},
```
Comment (lunaroyster) [inline modules/extensions/src/api/exec.ts:11]: i dont think we need onEnd, the promise itself should signal the end, no?

File: modules/extensions/src/util/comlink.ts [hunk 2]
```diff
@@ -1,12 +1,16 @@
...
   typeof window !== "undefined"
     ? (Comlink.wrap<ExtensionPortAPI>(
         Comlink.windowEndpoint(self.parent, self, "*")
-      ) as Comlink.Remote<ExtensionPortAPI>)
-    : null)() as Comlink.Remote<ExtensionPortAPI>;
+      ) as Comlink.Remote<ExtensionPortAPI> & {
+        experimental: Comlink.Remote<ExperimentalAPIs>;
+      })
+    : null)() as Comlink.Remote<ExtensionPortAPI> & {
+  experimental: Comlink.Remote<ExperimentalAPIs>;
```
Comment (lunaroyster) [inline modules/extensions/src/util/comlink.ts:12]: what errors do you see without this?

File: modules/extensions/src/api/exec.ts [hunk 3]
```diff
@@ -0,0 +1,59 @@
...
+      },
+    })
+  );
+
+  promise.then((res) => {
+    onEnd({
+      ...res,
+      output: outputStr,
+    });
+    dispose();
```
[the remark this answers is not in the document]
Comment (Conner1115) [reply modules/extensions/src/api/exec.ts:39]: Can remove, that's probably the case.

File: modules/extensions/src/api/exec.ts (hunk 1 above)
Comment (Conner1115) [reply modules/extensions/src/api/exec.ts:11]: The exec function returns a disposer while using a callback to signal for when it ends.  If we wait until the command finishes, there will be no way to stop a running process.

File: modules/extensions/src/util/comlink.ts (hunk 2 above)
Comment (Conner1115) [reply modules/extensions/src/util/comlink.ts:12]: Typescript errors.  Says I need to do `await (await extensionPort.experimental).exec(...)`, but that is incorrect.

File: modules/extensions/src/api/exec.ts [hunk 4]
```diff
@@ -0,0 +1,44 @@
...
+  onStdErr = () => {},
+  onError = () => {},
+  onEnd = () => {},
+}: ExecArgs): Promise<() => void> {
+  let outputStr: string = "";
+
+  // TODO: Make dispose() not cause a BSOD after being called immediately
+  const { dispose, promise } = await extensionPort.experimental.exec(
+    proxy({
+      args: Array.isArray(args) ? args : ["bash", "-c", args],
```
Comment (lunaroyster) [inline modules/extensions/src/api/exec.ts:18]: is `args` automatically escaped here?

if you enter `git commit -m "test"`, does that break due to the " or not?

File: modules/extensions/src/util/comlink.ts (hunk 2 above)
Comment (lunaroyster) [reply modules/extensions/src/util/comlink.ts:12]: hmm okay, i'll take a shot at something sweet and simple

**Review by lunaroyster — APPROVED** [review]
LGTM!

**Review by lunaroyster — APPROVED** [review]
LGTM!

File: modules/example/App.tsx [hunk 5]
```diff
@@ -1,16 +1,40 @@
...
   const { status, error, filePath, replit } = useReplit();
 
-  if (typeof window !== 'undefined') {
+  if (typeof window !== "undefined") {
     // @ts-ignore
     window.replit = replit;
   }
 
+  const execute = async () => {
+    const dispose = await exec({
```
Comment (lunaroyster) [inline modules/example/App.tsx:22]: actually you know what, i really don't like `await` here referring to the start of execution rather than the end

**Review by lunaroyster — CHANGES_REQUESTED** [review]
Let's focus on just returning a promise for the end of execution. Most people just want to run a command

We can always add an AbortSignal and timeout later

File: modules/extensions/src/api/exec.ts [hunk 6]
```diff
@@ -0,0 +1,66 @@
...
+  let errorStr: string = "";
+  let exitCode: string = "";
+
+  const { promise } = await extensionPort.experimental.exec(
+    proxy({
+      args: Array.isArray(options.args)
+        ? options.args
+        : ["bash", "-c", options.args],
+      env: options.env || {},
+      splitStderr: options.separateStdErr,
```
Comment (lunaroyster) [inline modules/extensions/src/api/exec.ts:28]:
```suggestion
      splitStderr: separateStdErr,
```

File: modules/extensions/src/api/exec.ts [hunk 7]
```diff
@@ -0,0 +1,66 @@
...
+      onOutput: (output: string) => {
+        outputStr += output;
+        if (options.separateStdErr) {
+          options.onStdOutOutput?.(output);
+        } else {
+          options.onOutput?.(output);
+        }
+      },
+      onStdErr: (stderr: string) => {
+        if (options.separateStdErr) {
```
Comment (lunaroyster) [inline modules/extensions/src/api/exec.ts:38]:
```suggestion
        if (splitStdErr) {
          options.onStdOutOutput?.(output);
        } else {
          options.onOutput?.(output);
        }
      },
      onStdErr: (stderr: string) => {
        if (splitStdErr) {
```

File: modules/extensions/src/api/exec.ts [hunk 8]
```diff
@@ -0,0 +1,66 @@
...
+  separatedOutputOptions: SeparatedOutputExecOptions
+): Promise<SeparatedOutputExecResult>;
+export async function exec(
+  options: CombinedOutputExecOptions | SeparatedOutputExecOptions
+): Promise<SeparatedOutputExecResult | CombinedOutputExecResult> {
+  let outputStr: string = "";
+  let errorStr: string = "";
+  let exitCode: string = "";
+
+  const { promise } = await extensionPort.experimental.exec(
```
[the remark this answers is not in the document]
Comment (lunaroyster) [reply modules/extensions/src/api/exec.ts:22]: oh we may also need to await that

[the remark this answers is not in the document]
Comment (Conner1115) [reply modules/extensions/src/api/exec.ts:24]: Why do we need to await options.separateStdErr?

[the remark this answers is not in the document]
Comment (Conner1115) [reply modules/extensions/src/api/exec.ts:24]: Typescript is not happy with this lol

**Review by lunaroyster — APPROVED** [review]
let's goooo!

**Pull request merged**